### PR TITLE
Explicitly set privileged istio-cni to false

### LINKED ISSUE #52558: Allow running istio-cni as unprivileged
**Describe the feature request**
Today, the main istio-cni container runs as a Daemonset with `privileged: true`. This isn't great to have a privileged container running on every node; instead, we should do whatever privileged actions are necessary in a separate initcontainer (this is the approach kube-proxy takes: https://www.kubernetes.dev/blog/2024/01/05/kube-proxy-non-privileged/). Note that the repair controller currently requires `privileged: true` because it reads procfs; we should allow turning that off if we don't already. Users may have other options (e.g. a cloud provider process) for handling the race-conditions, so istio-cni should be able to turn off that functionality

**Describe alternatives you've considered**

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[ ] Dual Stack
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

### PR DESCRIPTION
**Pull request by jaellio**
**Please provide a description of this PR:**

Set privileged to false and add capabilities based on feature. Remove unnecessary dependency on `privileged: true`

TODO: (items may be captured in unit tests)
- [ ] Verify functionality in openshift (Fixed with https://github.com/istio/istio/pull/53529)
- [x] Verify functionality of repair controller 
  - Requires `CAP_SYS_PTRACE`
- [x] Verify functionality in ambient mode.

### CODE REVIEW

File: manifests/charts/istio-cni/templates/daemonset.yaml [hunk 1]
```diff
@@ -89,18 +89,14 @@ spec:
...
+            privileged: false
             runAsGroup: 0
             runAsUser: 0
             runAsNonRoot: false
             # Both ambient and sidecar repair mode require elevated node privileges to function.
-            # But we don't need _everything_ in `privileged`, so drop+readd capabilities based on feature.
-            # privileged is redundant with CAP_SYS_ADMIN
-            # since it's redundant, hardcode it to `true`, then manually drop ALL + readd granular
-            # capabilities we actually require
+            # But we don't need _everything_ in `privileged`, so explicitly set it to false and
```
Comment (bleggett) [inline manifests/charts/istio-cni/templates/daemonset.yaml:97]: This is a regrettable feature of the `securityContext.privileged` API in K8S. It's a bad API. You can set it to `false` but still have root on the node - which is what we had before. Which was massively confusing to end-users.

We are self-granting `CAP_SYS_ADMIN` and running as root, which means this pod _is_ completely privileged as root on the node in actual fact, regardless of the value of `securityContext.privileged`.

This is why, for instance, [in the k8s Pod Security Standards, `securityContext.privileged` and `securityContext.capabilities` are always lockstepped](https://kubernetes.io/docs/concepts/security/pod-security-standards/#baseline)

As long as we have CAP_SYS_ADMIN `istio-cni` is absolutely a privileged container, any way you measure it, regardless of the value of `securityContext.privileged`.

And setting this to `false` while keeping the `capabilities.add.SYS_ADMIN` still doesn't allow us to run under a different PSS profile anyway - [we are still stuck with `privileged`](https://kubernetes.io/docs/concepts/security/pod-security-standards/#privileged)

So I don't think setting this to `false` actually does anything but mislead people and get us past not-very-smart manifest-based pod admission gates.

File: manifests/charts/istio-cni/templates/daemonset.yaml [hunk 2]
```diff
@@ -89,18 +89,14 @@ spec:
...
+            # add capabilities based on feature.
             capabilities:
-              drop:
-              - ALL
```
Comment (bleggett) [inline manifests/charts/istio-cni/templates/daemonset.yaml:103]: Why drop this? If we do not drop ALL then add back only what we need, I think the net effect is we have more perms by default?

(granted, since we still have CAP_SYS_ADMIN, it doesn't matter)

Comment (jaellio) [reply manifests/charts/istio-cni/templates/daemonset.yaml:103]: I don't think CAP_SYS_ADMIN is the same as privileged. Setting `securityContext.privileged=true`, then dropping all capabilities, and adding back NET_RAW, NET_ADMIN and SYS_ADMIN is not the same thing as setting `securityContext.privileged=false` and adding NET_RAW, NET_ADMIN and SYS_ADMIN. `securityContext.privileged=true` is more encompassing that CAP_SYS_ADMIN.

For example, from the failed integ-cni_istio tests, setting NET_RAW, NET_ADMIN and SYS_ADMIN was not sufficient for the repair controller. I needed to additionally set CAP_SYS_PTRACE to allow describing the pods net ns.

Comment (jaellio) [reply manifests/charts/istio-cni/templates/daemonset.yaml:103]: I am open to keeping the `securityContext.privileged` true by default but I'd like to give users the option to set `securityContext.privileged` to false if they need to do so for compliance reasons.

It is also concerning that setting `securityContext.privileged=true` means there are additional privileged beyond the capabilities we add after dropping all.

Comment (bleggett) [reply manifests/charts/istio-cni/templates/daemonset.yaml:103]: > I am open to keeping the `securityContext.privileged` true by default but I'd like to give users the option to set `securityContext.privileged` to false if they need to do so for compliance reasons.

If setting `securityContext.privileged=false`
- Doesn't make the pod non-privileged (it is still running as root, has total node privileges, and can grant itself any permissions)
- Still (because of the above) can't allow us to move out of the K8S `Privileged` PodSecurityStandard

then what would be the compliance reasons? It's like putting a sign on a door that says "locked", but not locking the door.

We did this before, and all it did was hopelessly confuse and mislead users. "I set `privileged` to false, so `istio-cni` isn't privileged, right"? -> no, it's still completely privileged, and runs as root on your node. There is no environment where `securityContext.privileged` is not an acceptable security risk, but running as root with CAP_SYS_ADMIN is. There are just people that misunderstand what `privileged` means.

> It is also concerning that setting `securityContext.privileged=true` means there are additional privileged beyond the capabilities we add after dropping all.

Yep, and that's a good catch. It looks like that's something in sidecar repair we missed. I am very much in favor of explicitly listing granular permissions and not relying on ~whatever mishmash of `CAP_XX` privileges `securityContext.privileged` happens to grant - it's a bad API and a vestige of the Docker days.

File: manifests/charts/istio-cni/templates/daemonset.yaml (hunk 1 above)
Comment (jaellio) [reply manifests/charts/istio-cni/templates/daemonset.yaml:97]: I totally agree setting privileged to false is confusing and there is little difference between the options when setting CAP_SYS_ADMIN.

My main motivation is allowing user to explicitly see the permissions being added and be able to evaluate the risk of each. Setting privileged to true and then dropping all capabilities still leaves elevated privileges (in addition to the capabilities we currently add back). I think this scenario is also confusing to the user.

File: manifests/charts/istio-cni/templates/daemonset.yaml (hunk 2 above)
Comment (jaellio) [reply manifests/charts/istio-cni/templates/daemonset.yaml:103]: To be clear, I totally agree that setting `securityContext.privileged=false` does not mean that the pod isn't privileged. And I agree that users not wanting `securityContext.privileged=true` for compliance reasons but are okay with CAP_SYS_ADMIN means they either want to check a compliance box or they misunderstand what privileged means. In this scenario I think both reasoning are valid given top down compliance requirements and the complexed of the API.

Comment (jaellio) [reply manifests/charts/istio-cni/templates/daemonset.yaml:103]: > Yep, and that's a good catch. It looks like that's something in sidecar repair we missed. I am very much in favor of explicitly listing granular permissions under securityContext.capabilities and not relying on ~whatever mishmash of CAP_XX privileges securityContext.privileged happens to grant under whatever container runtime you happen to be using - it's a bad API and a vestige of the Docker days.

My concern is we'll only catch these instances when `securityContext.privileged` is `false`.

Comment (jaellio) [reply manifests/charts/istio-cni/templates/daemonset.yaml:103]: Is it preferable to explicitly enumerate capabilities and guarantee there are no hidden privileges while potentially introducing confusion by setting `securityContext.privileged` to `false` OR is better to set `securityContext.privileged` to `true` and clearly communicate to the user that the container has privileged capabilities which potentially allowing unknown container privileges?

Comment (jaellio) [reply manifests/charts/istio-cni/templates/daemonset.yaml:103]: > Still (because of the above) can't allow us to move out of the K8S Privileged PodSecurityStandard

Could you explain more about this? Not super familiar with PSS

File: manifests/charts/istio-cni/templates/daemonset.yaml (hunk 1 above)
Comment (bleggett) [reply manifests/charts/istio-cni/templates/daemonset.yaml:97]: > Setting privileged to true and then dropping all capabilities still leaves elevated privileges (in addition to the capabilities we currently add back). I think this scenario is also confusing to the user.

+1 on this, it's a good point and I don't object to hardcoding it to `false`. 

In general, I want to hew to the published K8S Pod Security Standards  - as per those, this pod is _always_ in the Privileged profile, and _never_ in the Baseline or Restricted profile.

I only object to making it a toggle to fit an unspecified "compliance" profile that isn't the K8S Pod Security Standard - because any compliance policy that would reject `securityContext.privileged: true` but not reject `securityContext.capabilities.CAP_SYS_ADMIN` is explicitly broken from the perspective of that standard.

File: manifests/charts/istio-cni/templates/daemonset.yaml (hunk 2 above)
Comment (bleggett) [reply manifests/charts/istio-cni/templates/daemonset.yaml:103]: Yeah - if you look at [the doc ](https://kubernetes.io/docs/concepts/security/pod-security-standards/#baseline)it has better detail than I can give, but roughly:

1. There are (currently) three pod profiles, ranging from least-privileged to most: Restricted, Baseline, Privileged

2. Those three profiles are defined according to what the pod may or may not declare as privileges for itself. .e.g for a pod to be in Baseline or Restricted, among other things

- permittted `securityContext.capabilities.add` capabilities are allowlisted (i.e. no `CAP_SYS_ADMIN`)
- no host path volumes
- etc

Basically it's an attempt to fix the bad/insufficiently granular API that is `securityContext.privileged: <YES/NO>` - any pod that can't meet Restricted (the most locked down) gets bumped up to Baseline (which is the explicitly recommended default at this time for all pods), and anything that can't meet Baseline is Privileged.

Once your pod is in the PSS Privileged category, it's just Privileged. There's no additional granularity, or Semi-Privileged profile (if they add something like that, I'd be happy to put us in the lowest one we can fit into).

File: manifests/charts/istio-cni/templates/daemonset.yaml (hunk 1 above)
Comment (jaellio) [reply manifests/charts/istio-cni/templates/daemonset.yaml:97]: > I only object to making it an exposed toggle to fit an unspecified "compliance" profile that isn't the K8S Pod Security Standard - because any compliance policy that would reject securityContext.privileged: true but not reject securityContext.capabilities.CAP_SYS_ADMIN is explicitly broken from the perspective of that standard.

+1

**Review by bleggett — APPROVED** [review]
LGTM.

File: manifests/charts/istio-cni/templates/daemonset.yaml [hunk 3]
```diff
@@ -89,26 +89,25 @@ spec:
...
             runAsUser: 0
             runAsNonRoot: false
             # Both ambient and sidecar repair mode require elevated node privileges to function.
-            # But we don't need _everything_ in `privileged`, so drop+readd capabilities based on feature.
-            # privileged is redundant with CAP_SYS_ADMIN
-            # since it's redundant, hardcode it to `true`, then manually drop ALL + readd granular
-            # capabilities we actually require
+            # But we don't need _everything_ in `privileged`, so explicitly set it to false and
+            # add capabilities based on feature.
             capabilities:
```
Comment (howardjohn) [inline manifests/charts/istio-cni/templates/daemonset.yaml:99]: Asking the same question in https://github.com/istio/istio/pull/53478/files#r1792582294 since it didn't seem to get directly answered -- why we do not `drop: ALL`? Can we still?

Comment (bleggett) [reply manifests/charts/istio-cni/templates/daemonset.yaml:99]: Yeah - good call I missed that this go-around. We should `drop: ALL` as a matter of course

Comment (jaellio) [reply manifests/charts/istio-cni/templates/daemonset.yaml:99]: Agree, updated

**Review by howardjohn — APPROVED** [review]

**Comment by howardjohn** [comment]
Nice work!

**Comment by dgn** [comment]
FYI this commit breaks istio-cni on OpenShift. Might be due to SELinux. We're investigating if we can make it work without `privileged: true`

**Comment by keithmattix** [comment]
Can we change it back when the open shift profile is enabled?

**Comment by bleggett** [comment]
> Update: actually it looks like it's just the fact that on OpenShift, you can't use `hostPath` without `privileged: true`.

Is this something we can do by adding the right CAP_XXX?

If we can't, it might imply that OpenShift is correct, and `privileged: true` is required (we do unequivocally require hostmounts - and it might simply work elsewhere due to lax enforcement of hostmount capabilities.

**Comment by keithmattix** [comment]
According to the [docs](https://kubernetes.io/docs/concepts/storage/volumes/#hostpath), privileged is only required if the   path requires root perms

**Comment by luksa** [comment]
@bleggett this can't be fixed by adding any `CAP_` because it's caused by SELinux and the fact that the file is under `/var/run` (and thus has the `container_var_run_t` type, which the container itself does not). 

This can be set by:
```
securityContext:
  seLinuxOptions:
    type: spc_t
```

Interestingly, this value is already set in `profile-openshift.yaml`, but this value is only used in the ztunnel chart. Maybe we just need to use the same value in the CNI chart?

**Comment by luksa** [comment]
PR: https://github.com/istio/istio/pull/53529
